pkg/reconciler/ingress/lister.go

```diff
@@ -0,0 +1,176 @@
+func NewStatusProber(
+	logger *zap.SugaredLogger,
+	gatewayLister istiolisters.GatewayLister,
+	endpointsLister corev1listers.EndpointsLister,
+	serviceLister corev1listers.ServiceLister,
+	readyCallback func(*v1alpha1.Ingress)) *status.Prober {
+	return status.NewProber(
+		logger,
+		&gatewayPodTargetLister{
+			logger:          logger,
+			gatewayLister:   gatewayLister,
+			endpointsLister: endpointsLister,
+			serviceLister:   serviceLister,
+		},
+		readyCallback,
+	)
+}
+
+type gatewayPodTargetLister struct {
+	logger *zap.SugaredLogger
+
+	gatewayLister   istiolisters.GatewayLister
+	endpointsLister corev1listers.EndpointsLister
+	serviceLister   corev1listers.ServiceLister
+}
+
+// probeTargets represents the target to probe.
+type probeTarget struct {
+	// urlTmpl is an url template to probe.
+	urlTmpl string
+	// targetPort is a port of Gateway pod.
+	targetPort string
+}
+
+func (l *gatewayPodTargetLister) ListProbeTargets(ctx context.Context, ing *v1alpha1.Ingress) (map[string]map[string]sets.String, error) {
+	results := map[string]map[string]sets.String{}
+
+	qualifiedGatewayNames := qualifiedGatewayNamesFromContext(ctx)
+
+	gatewayHosts := ingress.HostsPerVisibility(ing, qualifiedGatewayNames)
+	for gatewayName, hosts := range gatewayHosts {
+		gateway, err := l.getGateway(gatewayName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get Gateway %q: %w", gatewayName, err)
+		}
+		targetsPerPod, err := l.listGatewayTargetsPerPods(gateway)
+		if err != nil {
+			return nil, fmt.Errorf("failed to list the probing URLs of Gateway %q: %w", gatewayName, err)
+		}
+		if len(targetsPerPod) == 0 {
+			continue
+		}
+		for ip, targets := range targetsPerPod {
+			if _, existed := results[ip]; !existed {
+				results[ip] = map[string]sets.String{}
+			}
+			for _, host := range hosts.List() {
+				for _, target := range targets {
+					url := fmt.Sprintf(target.urlTmpl, host)
+
+					if targets, existed := results[ip][target.targetPort]; existed {
+						targets.Insert(url)
+					} else {
+						results[ip][target.targetPort] = sets.NewString(url)
+					}
+				}
+			}
+		}
+	}
+	return results, nil
+}
+
+func (l *gatewayPodTargetLister) getGateway(name string) (*v1alpha3.Gateway, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse Gateway name %q: %w", name, err)
+	}
+	if namespace == "" {
+		return nil, fmt.Errorf("unexpected unqualified Gateway name %q", name)
+	}
+	return l.gatewayLister.Gateways(namespace).Get(name)
+}
+
+// listGatewayPodsURLs returns a map where the keys are the Gateway Pod IPs and the values are the corresponding
+// URL templates and the Gateway Pod Port to be probed.
+func (l *gatewayPodTargetLister) listGatewayTargetsPerPods(gateway *v1alpha3.Gateway) (map[string][]probeTarget, error) {
+	selector := labels.SelectorFromSet(gateway.Spec.Selector)
+
+	services, err := l.serviceLister.List(selector)
```

Comment:
Could we just list the endpoints directly? Wouldn't they have the same labels?

Reply:
This is a great suggestion.  I will follow up, since I mostly keep the existing behavior in this version.

Reply:
You can't do this because of the need to resolve the port name through the Service.

Reply:
(I tried in my initial version)